### ISSUE TITLE
Building lua-nginx-module on RaspberryPi with LuaJIT 2.0.5 -> 32bit size problem

### ISSUE THREAD

**Issue by alecmuffett** [openresty/lua-nginx-module#1423]
Ensure you have provided the following details while reporting a problem:

* The exact version of the related software, including but not limited to the OpenResty version
(if any), the NGINX core version, the `ngx_lua` module version,
and your operating system version.

I am running Raspbian Stretch (a 32-bit operating system) latest, fully updated, on a Raspberry Pi 3b.

* A minimal and standalone test case that others can easily run on their side and
reproduce the issue you are seeing.

That's harder to offer, but I am simplu trying to build NGINX with the configure line:

```
+ env ./configure --with-cc-opt=-fPIE -fstack-protector-strong -fexceptions -D_FORTIFY_SOURCE=2 --with-ld-opt=-Wl,-Bsymbolic-functions -fPIE -pie -Wl,-z,relro -Wl,-z,now -Wl,-rpath,/big/home/pi/eotk/opt.d/usr/local/lib --prefix=/big/home/pi/eotk/opt.d --with-http_sub_module --with-http_ssl_module --add-module=/big/home/pi/eotk/opt.d/nginx-1.14.1/headers-more-nginx-module --add-module=/big/home/pi/eotk/opt.d/nginx-1.14.1/lua-nginx-module --add-module=/big/home/pi/eotk/opt.d/nginx-1.14.1/ngx_devel_kit --add-module=/big/home/pi/eotk/opt.d/nginx-1.14.1/ngx_http_substitutions_filter_module
```

...and the following additional modules:

```
configuring additional modules
adding module in /big/home/pi/eotk/opt.d/nginx-1.14.1/headers-more-nginx-module
 + ngx_http_headers_more_filter_module was configured
adding module in /big/home/pi/eotk/opt.d/nginx-1.14.1/lua-nginx-module
checking for LuaJIT library in /big/home/pi/eotk/opt.d/usr/local/lib and /big/home/pi/eotk/opt.d/usr/local/include/luajit-2.0 (specified by the LUAJIT_LIB and LUAJIT_INC env, with -ldl) ... found
checking for export symbols by default (-E) ... found
checking for export symbols by default (--export-all-symbols) ... not found
checking for SO_PASSCRED ... found
checking for SA_RESTART ... found
checking for __attribute__(constructor) ... found
checking for malloc_trim ... found
 + ngx_http_lua_module was configured
adding module in /big/home/pi/eotk/opt.d/nginx-1.14.1/ngx_devel_kit
 + ngx_devel_kit was configured
adding module in /big/home/pi/eotk/opt.d/nginx-1.14.1/ngx_http_substitutions_filter_module
 + ngx_http_subs_filter_module was configured
checking for PCRE library ... found
checking for PCRE JIT support ... found
checking for OpenSSL library ... found
checking for zlib library ... found
creating objs/Makefile
```

* Do not simply say "something is broken" or "something does not work". Always provide
as much details as possible. Always describe the symptoms and your expected results.

It was building previously but seems to have suddenly stopped; the build is now crashing out, reporting a shift operation which would be illegal on a 32-bit integer:

```
cc -c -I/big/home/pi/eotk/opt.d/usr/local/include/luajit-2.0  -pipe  -O -W -Wall -Wpointer-arith -Wno-unused-parameter -Werror -g -fPIE -fstack-protector-strong -fexceptions -D_FORTIFY_SOURCE=2 -DNDK_SET_VAR  -I src/core -I src/event -I src/event/modules -I src/os/unix -I /big/home/pi/eotk/opt.d/nginx-1.14.1/lua-nginx-module/src/api -I /big/home/pi/eotk/opt.d/nginx-1.14.1/ngx_devel_kit/objs -I objs/addon/ndk -I objs -I src/http -I src/http/modules -I /big/home/pi/eotk/opt.d/usr/local/include/luajit-2.0 -I /big/home/pi/eotk/opt.d/nginx-1.14.1/ngx_devel_kit/src -I /big/home/pi/eotk/opt.d/nginx-1.14.1/ngx_devel_kit/src -I /big/home/pi/eotk/opt.d/nginx-1.14.1/ngx_devel_kit/objs -I objs/addon/ndk \
        -o objs/addon/src/ngx_http_lua_subrequest.o \
        /big/home/pi/eotk/opt.d/nginx-1.14.1/lua-nginx-module/src/ngx_http_lua_subrequest.c
cc -c -I/big/home/pi/eotk/opt.d/usr/local/include/luajit-2.0  -pipe  -O -W -Wall -Wpointer-arith -Wno-unused-parameter -Werror -g -fPIE -fstack-protector-strong -fexceptions -D_FORTIFY_SOURCE=2 -DNDK_SET_VAR  -I src/core -I src/event -I src/event/modules -I src/os/unix -I /big/home/pi/eotk/opt.d/nginx-1.14.1/lua-nginx-module/src/api -I /big/home/pi/eotk/opt.d/nginx-1.14.1/ngx_devel_kit/objs -I objs/addon/ndk -I objs -I src/http -I src/http/modules -I /big/home/pi/eotk/opt.d/usr/local/include/luajit-2.0 -I /big/home/pi/eotk/opt.d/nginx-1.14.1/ngx_devel_kit/src -I /big/home/pi/eotk/opt.d/nginx-1.14.1/ngx_devel_kit/src -I /big/home/pi/eotk/opt.d/nginx-1.14.1/ngx_devel_kit/objs -I objs/addon/ndk \
        -o objs/addon/src/ngx_http_lua_ndk.o \
        /big/home/pi/eotk/opt.d/nginx-1.14.1/lua-nginx-module/src/ngx_http_lua_ndk.c
cc -c -I/big/home/pi/eotk/opt.d/usr/local/include/luajit-2.0  -pipe  -O -W -Wall -Wpointer-arith -Wno-unused-parameter -Werror -g -fPIE -fstack-protector-strong -fexceptions -D_FORTIFY_SOURCE=2 -DNDK_SET_VAR  -I src/core -I src/event -I src/event/modules -I src/os/unix -I /big/home/pi/eotk/opt.d/nginx-1.14.1/lua-nginx-module/src/api -I /big/home/pi/eotk/opt.d/nginx-1.14.1/ngx_devel_kit/objs -I objs/addon/ndk -I objs -I src/http -I src/http/modules -I /big/home/pi/eotk/opt.d/usr/local/include/luajit-2.0 -I /big/home/pi/eotk/opt.d/nginx-1.14.1/ngx_devel_kit/src -I /big/home/pi/eotk/opt.d/nginx-1.14.1/ngx_devel_kit/src -I /big/home/pi/eotk/opt.d/nginx-1.14.1/ngx_devel_kit/objs -I objs/addon/ndk \
        -o objs/addon/src/ngx_http_lua_control.o \
        /big/home/pi/eotk/opt.d/nginx-1.14.1/lua-nginx-module/src/ngx_http_lua_control.c
In file included from /big/home/pi/eotk/opt.d/nginx-1.14.1/lua-nginx-module/src/ngx_http_lua_control.h:12:0,
                 from /big/home/pi/eotk/opt.d/nginx-1.14.1/lua-nginx-module/src/ngx_http_lua_control.c:14:
/big/home/pi/eotk/opt.d/nginx-1.14.1/lua-nginx-module/src/ngx_http_lua_control.c: In function ‘ngx_http_lua_on_abort’:
/big/home/pi/eotk/opt.d/nginx-1.14.1/lua-nginx-module/src/ngx_http_lua_common.h:144:61: error: left shift count >= width of type [-Werror=shift-count-overflow]
     ((void *) ((uintptr_t) (&ngx_http_lua_##ludata) & ((1UL << 47) - 1)))
                                                             ^
/big/home/pi/eotk/opt.d/nginx-1.14.1/lua-nginx-module/src/ngx_http_lua_control.c:435:30: note: in expansion of macro ‘ngx_http_lua_lightudata_mask’
     lua_pushlightuserdata(L, ngx_http_lua_lightudata_mask(
                              ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
cc1: all warnings being treated as errors
objs/Makefile:1421: recipe for target 'objs/addon/src/ngx_http_lua_control.o' failed
make[1]: *** [objs/addon/src/ngx_http_lua_control.o] Error 1
make[1]: Leaving directory '/big/home/pi/eotk/opt.d/nginx-1.14.1'
Makefile:8: recipe for target 'build' failed
make: *** [build] Error 2
```

**Comment by alecmuffett** [comment]
Perhaps when @doujiang24  added support for ARM64, he may have broken ARM32?

https://github.com/openresty/lua-nginx-module/commit/7286812116940216344ade33722c49ae47037605#diff-198bce3a7eb5e4a6fdc54aae29b46483L435

**Comment by alecmuffett** [comment]
Confirmed, if I roll back to `f64ec8c9056437bb1b822fc4e99f05f784d36e32` the code compiled cleanly.  @doujiang24, why is there a hardcoded 47-bit mask-off, please? 

Presumably some kind of tagging is being done in the higher bits, but does it really need to happen on all architectures, please? RaspberryPi is popular, and ARM32 still exists...

**Comment by thibaultcha** [comment]
Nice catch @alecmuffett. On mobile atm, but @spacewander and @doujiang24 had a look and said they’ll try to fix it before the next release. In the meantime I suggest that you stick to the latest non-release-candidate version of this module.

**Comment by thibaultcha** [comment]
@alecmuffett Could you confirm that @spacewander's fix addresses your issue? Waiting on your reply before merging. Thank you.

**Comment by alecmuffett** [comment]
Alas, I can't work on this until Friday. More then.

**Comment by alecmuffett** [comment]
Confirmed. Compilation is successful with the patch applied.

**Comment by thibaultcha** [comment]
That’s great news. Thanks for getting back to us.

**Comment by thibaultcha** [comment]
Just merged #1424. Thanks again @alecmuffett :)
